### PR TITLE
[barelib] vsnprintf()

### PR DESCRIPTION
**Pull request by JoseGuilhermeCR**
# Description
Added a __vsnprintf function that will write at most n bytes (including the null terminating byte ('\0')) to a given buffer. Other than that the function offers the same functionality that __vsprintf does.

### CODE REVIEW

File: src/vsprintf.c [hunk 1]
```diff
@@ -200,6 +201,71 @@ static int itoa64(char *str, unsigned long long num, int base)
...
+#else
+		unsigned long long quocient = 0;
+		unsigned long long dividend = num;
+		while (dividend >= 10)
+		{
+			dividend -= 10;
+			quocient++;
+		}
+
+		num = quocient;
```
Comment (ppenna) [inline src/vsprintf.c:260]: Couldn't we use a macro for this that would expand to the appropriate type?

File: src/vsprintf.c [hunk 2]
```diff
@@ -213,33 +279,28 @@ int __vsprintf(char *str, const char *fmt, va_list args)
...
 					*str++ = (char)va_arg(args, int);
 					break;
-
 				/* Number. */
 				case 'd':
-					str += itoa(str, va_arg(args, unsigned int), 'd');
-					break;
 				case 'x':
-					str += itoa(str, va_arg(args, unsigned int), 'x');
+					str += itoa(str, va_arg(args, unsigned int), *fmt);
```
Comment (ppenna) [inline src/vsprintf.c:301]: Why change this?

File: src/vsprintf.c [hunk 3]
```diff
@@ -249,22 +310,114 @@ int __vsprintf(char *str, const char *fmt, va_list args)
...
-    }
+		++fmt;
+	}
 
-    return (str - base);
+	*str = '\0';
+
+	return ((str - base) - 1);
 }
+
```
Comment (ppenna) [inline src/vsprintf.c:423]: Why not use this function as a back end for `vsprintf()`? Or even better: if we do not use `vsprintf()` elsewhere, why don't we drop it?

Comment (JoseGuilhermeCR) [reply src/vsprintf.c:423]: On the barelib, sprintf is actually a wrapper that uses vsprintf. If I were to remove the original function, I'd have to change that, which is not a problem. But I don't know if code in other repositories use the function as well, so I think it's better to have vsnprintf as a backend. In that case, we would have to give it a max size, what would be the ideal value?

File: src/vsprintf.c (hunk 1 above)
Comment (JoseGuilhermeCR) [reply src/vsprintf.c:260]: Something like
```
#ifndef __qemu_riscv32__
#define BIGGEST_UNSIGNED unsigned long long
#else
#define BIGGEST_UNSIGNED unsigned
#endif
```
that would make it possible to reduce those duplicated functions into one?

Comment (ppenna) [reply src/vsprintf.c:260]: Exactly ;-)

File: src/vsprintf.c (hunk 3 above)
Comment (ppenna) [reply src/vsprintf.c:423]: I think that we should really drop it, just grep it to make sure where it is used. For the size I would say that you can use a macro with 128 bytes

**Pull request merged**